Feature/enumerable methods

SigmaSam opened the pull request:
- Finished all methods
- Made test for all scenarios from the documentation which they passed all.

Live review: https://repl.it/@SigmaSam/Enumerable-Methods#main.rb

Commit:
Create main.rb && Add my_each method.

my_each prints sample message

Commit:
Add class Arreglo to include Enumerable module and Fix Rubocop Alert

*Rubocop alerts with:
  module_function :my_each

Commit:
Remove classes & each replicated on my_each. Assignment 2.3 finished

each definition:
Calls the given block once for each element in self, passing that element as a parameter.
Returns the array itself

Commit:
Add test_hash to use with my_each_with_index

SigmaSam committed:
added select and each_with_index.

Commit:
Fix Rubocop issues && Add ternary operator to my_select

SigmaSam committed:
Finished any and none methods

Commit:
Add my_map. Fix counters on methods. Change testcases for my_count

Commit:
Fix my_map (change block on yield condition). Create my_inject method (not finished)

Commit:
Fix my_map, add yield result to new array or false. Change test to test_string

Commit:
Fix my_each_with_index: Add index parameter to yield block

Commit:
Fix redundant self on my_inject: Change self to i in yield parameters

Commit:
Fix my_count and my_all redundant self

SigmaSam committed:
Set work evioroment for testing methods.

Commit:
Working on issue #2

Commit:
FIX: my_each method finished

SigmaSam committed:
fixed method for testing.

Commit:
my_each finished. Create all test cases

Commit:
Add variables to tests.rb

Commit:
my_each_with_index finished.

Commit:
Add my_select method. *error on hashs

Commit:
Fix tests.rb & seudo.rb rubocop issues

SigmaSam committed:
fixed select for all but hashes.

Commit:
my_select method finished. All tests passed

SigmaSam committed:
aparently my_all? works now.

SigmaSam committed:
Merge remote-tracking branch 'origin/feature/my_all' into feature/my_all

Commit:
Working on my_all?, not working

Commit:
Change tests.rb to add the appropiate test cases for my_all

Commit:
Fix my_all on word.length conditional. Fix test cases

SigmaSam committed:
Merge remote-tracking branch 'origin/feature/my_all' into feature/my_all

SigmaSam committed:
Merge remote-tracking branch 'origin/feature/my_all' into feature/my_all

SigmaSam committed:
Build the flow logic for my_all

SigmaSam committed:
Personalizations in the test function.

Commit:
Change test cases and deleted comments

Commit:
Change test cases and deleted comments

SigmaSam committed:
Merge branch 'feature/my_all' of https://github.com/marcoshdezcam/enumerable_methods into feature/my_all

SigmaSam committed:
removed conflict.

SigmaSam committed:
Finished the logic for my_all

Commit:
Fix logic on my_all

SigmaSam committed:
Fixed test cases for working methods.

Commit:
Fix my_all: Decrease complexity and add all test cases

* Compares strings but cannot compare if arguments are all Numeric, for
  example
* Needs a fix to work on Hashes

Commit:
FIX my_all. Cannot cover all test cases with lower complexity

SigmaSam committed:
Started my none.

SigmaSam committed:
My_any works with all but arguments.

Commit:
Working on my_each

Commit:
my_all : Transforms to array Hashes and Ranges

Commit:
Start my_any method

SigmaSam committed:
My_count done.

Commit:
Method my_none? finished

Commit:
my_map method DONE

SigmaSam committed:
My_inject ver 0.1

SigmaSam committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/enumerable_methods'

Conflicts:
	tests.rb

Commit:
my_inject method DONE

*Doesn't work with last test case, FIX!

SigmaSam committed:
Missing multiply_els

SigmaSam committed:
my_map with proc

Commit:
FIX my_each syntax oon my_map

Commit:
multiply_els methhod DONE

Commit:
FIX my_inject method. *Needs work

Error on Block with initial value on accumulator

SigmaSam committed:
Finished modules.

SigmaSam committed:
added test file to gitignore.

SigmaSam committed:
Removed tests.rb file.

SigmaSam committed:
added Live version.

SigmaSam committed:
Modified gitignore

SigmaSam committed:
Removed rubocop warnings.

SigmaSam committed:
Bundle config update.

Commit:
Add flow control for symbols on the arguments

SigmaSam committed:
Removed Gemfiles because conflicts with the rspec

lekegitrepo commented:
#### Hi @SigmaSam and @marcoshdezcam :clap: :clap:
#### You have done a good job so far. However, some of the requirements are missing. check my comment below :point_down:

**_1. You are supposed to work with Gitflow in this project._**

**_2. Remove the unused file (seudo.rb) file._**

**_3. `my_each_with_index`:_**
**_`(1..5).my_each_with_index { |element, index| puts "#{element} => #{index}" }`_**
**_should return:_** 

**_`1 => 0`_**
**_`2 => 1`_**
**_`3 => 2`_**
**_`4 => 3`_**
**_`5 => 4`_**

**_`[1, 2, 3, 4, 5].my_each_with_index { |element, index| puts "#{element} => #{index}" }`_**

**_`1 => 0`_**
**_`2 => 1`_**
**_`3 => 2`_**
**_`4 => 3`_**
**_`5 => 4`_**

**_4. my_all?:_**
**_`[1, 5i, 5.67].my_all?(Numeric) # => true`_**
**_`[2, 1, 6, 7, 4, 8, 10].my_all?(Integer) # => true`_**

**_5. my_any?:_**
**_`[1, 5i, 5.67].my_any?(Numeric) # => true`_**
**_`[2, 1, 6, 7, 4, 8, 10].my_any?(Integer) # => true`_**

**_6. my_none?:_**
**_`[1, 5i, 5.67].my_any?(Numeric) # => false`_**
**_`[1, 3.14, 42].my_none?(Float) # => false`_**

**_implement the required changes and submit them for review._**

#### _Reviewed by Temitayo Adeleke._

lekegitrepo reviewed (changes requested):
#### Status: Request changes

Commit:
FIX: Delete seudo.rb && Solve my_each_with_index method

SigmaSam committed:
Completed my all? to work with hashes and ranges.

Commit:
Add tests.rb to .gitignore && Fix my_all? method with arguments

Commit:
Delete spaces on my_all?

Commit:
Solve Rubocop issue on my_all?

SigmaSam committed:
Removed rubocop erros with my_all?

SigmaSam committed:
Fixed my_any to take class arguments.

Commit:
my_none method finished. *Applied changes suggested by TSE

marcoshdezcam commented:
Changes requested by TSE done: 

1. Change pull request to follow Gitflow
2. Delete seudo.rb
3. Fix each_with_index to work with arguments
4. Fix my_all to work with hashes and ranges && Fix linter errors
5. Fix my_any to take class arguments
6. Fix my_none: change comparison with arguments

SigmaSam committed:
Fixed my all to return false when it has no block or argument.

christianotieno commented on main.rb:
```diff
@@ -0,0 +1,158 @@
+    item
+  end
+
+  def my_all?(*args)
+    flag = 0
+    i = 0
+    arr = self
+    is_a?(Range) || is_a?(Hash) ? arr = arr.to_a : nil
+
+    case args.empty?
+    when true
+      if block_given?
+        blk = yield(arr.at(i))
+        arr.my_each { flag += 1 if blk == true }
+      else
+        arr.my_each { flag += 1 unless i.nil? || i == false }
+      end
+    when false
+      arr.my_each { flag += 1 if i.is_a?(args[0]) }
+    end
+    return true if flag == size
+    return false if flag != size
+  end
```
# my_all? 
1. [x] When no block or argument is given returns false when one of the collection members is false or nil

     Failure/Error: expect(false_array.my_all?).to be false_array.all?

       expected false
            got true

##
# my_all?
2. [x] When a class is passed as an argument returns false if any of the collection is not a member of such class

     Failure/Error: expect(array.my_all?(Integer)).to be array.all?(Integer)

       expected false
            got true
##

# my_all? 
3. [x] When a Regex is passed as an argument returns true if all of the collection matches the Regex

     Failure/Error: expect(words.my_all?(/d/)).to be words.all?(/d/)

     TypeError:
       class or module required
##
# my_all? 
4. [x] When a Regex is passed as an argument returns false if any of the collection does not match the Regex

     Failure/Error: expect(words.my_all?(/o/)).to be words.all?(/o/)

     TypeError:
       class or module required
##

# my_all? 
5. [x] When a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

     Failure/Error: expect(array.my_all?(3)).to be array.all?(3)

     TypeError:
       class or module required
##     
# my_all? 
6. [x] When a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern

     Failure/Error: expect(array.my_all?(3)).to be array.all?(3)

     TypeError:
       class or module required

christianotieno commented on main.rb:
```diff
@@ -0,0 +1,158 @@
+    return false if flag != size
+  end
+
+  def my_any?(*args)
+    flag = 0
+    case args.empty?
+    when true
+      if block_given?
+        my_each { |i| flag += 1 if yield(i) }
+      else
+        my_each { |i| flag += 1 unless i.nil? || i == false }
+      end
+    when false
+      my_each { |i| flag += 1 if i.is_a?(args[0]) }
+    end
+    return true if flag.positive?
+    return false if flag <= 0
+  end
```
# my_any? 
1. [x] When a Regex is passed as an argument returns true if any of the collection matches the Regex

     Failure/Error: expect(words.my_any?(/d/)).to be words.any?(/d/)

     TypeError:
       class or module required
##
# my_any? 
2. [x] When a Regex is passed as an argument returns false if none of the collection matches the Regex

     Failure/Error: expect(words.my_any?(/z/)).to be words.any?(/z/)

     TypeError:
       class or module required
##
# my_any? 
3. [x] When a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

     Failure/Error: expect(words.my_any?('cat')).to be words.any?('cat')

     TypeError:
       class or module required
##   
# my_any? 
4. [x] When a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern

      Failure/Error: expect(words.my_any?('cat')).to be words.any?('cat')

      TypeError:
        class or module required

christianotieno commented on main.rb:
```diff
@@ -0,0 +1,158 @@
+    return false if flag <= 0
+  end
+
+  def my_none?(*args)
+    counter = 0
+    case !block_given?
+    when true
+      if args.empty?
+        my_each { |i| counter += 1 if i != true }
+      else
+        my_each { |i| counter += 1 if i == args[0] || i.is_a?(args[0]) }
+      end
+    when false
+      my_each { |i| counter += 1 if yield(i) == true }
+    end
+    return true if counter.zero?
+    return false if counter.positive?
+  end
```
# my_none? 
1. [x] When no block or argument is given returns true only if none of the collection members is true

      Failure/Error: expect(false_array.my_none?).to be true

        expected true
             got false
##
# my_none? 
2. [x] When a Regex is passed as an argument returns true if none of the collection matches the Regex

      Failure/Error: expect(words.my_none?(/z/)).to be words.none?(/z/)

      TypeError:
        class or module required
##
# my_none? 
3. [x] When a Regex is passed as an argument returns false if any of the collection matches the Regex

      Failure/Error: expect(words.my_none?(/d/)).to be words.none?(/d/)

      TypeError:
        class or module required
##
# my_none? 
4. [x] When a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

      Failure/Error: expect(words.my_none?(5)).to be words.none?(5)

      TypeError:
        class or module required
##    
# my_none? 
5. [x] When a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern

      Failure/Error: expect(words.my_none?(5)).to be words.none?(5)

      TypeError:
        class or module required

christianotieno commented on main.rb:
```diff
@@ -0,0 +1,158 @@
+    end
+  end
+
+  def my_map(*arg)
+    case block_given?
+    when true
+      new_array = []
+      my_each do |i|
+        new_array << yield(i)
+      end
+      new_array
+    when false && !arg.empty?
+      proc { arg }
+    when false && arg.empty?
+      return to_enum(:my_map) unless block_given?
+    end
+  end
```
# my_map 
- [x] Returns an Enumerator if no block is given

      Failure/Error: expect(array.my_map).to be_an(Enumerator)
##

# my_map? 
- [ ] Executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.

      Failure/Error: expect(array.my_map(my_proc) { |num| num < 10 }).to eq array.map(&my_proc)

        expected: [false, false, false, false, false, false, false, false, false, false, false, false, false, false, fa...e, false, false, false, false, false, false, false, false, false, false, false, false, false, false]
             got: [true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true...rue, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true]

christianotieno commented on main.rb:
```diff
@@ -0,0 +1,158 @@
+    end
+  end
+
+  def my_inject(arg = nil, sim = nil)
+    if block_given?
+      acc = arg
+      my_each { |i| acc = acc.nil? ? i : yield(acc, i) }
+    elsif !arg.nil? && arg.is_a?(Symbol)
+      acc = nil
+      my_each { |i| acc = acc.nil? ? i : acc.send(arg, i) }
+    elsif !sim.nil? && sim.is_a?(Symbol)
+      acc = arg
+      my_each { |i| acc = acc.nil? ? i : acc.send(sim, i) }
+    end
+    acc
+  end
+end
```
# my_inject 
- [x] Raises a "LocalJumpError" when no block or argument is given

      Failure/Error: expect { array.my_inject }.to raise_error(LocalJumpError)
        expected LocalJumpError but nothing was raised

christianotieno reviewed (changes requested):
# **CHANGES REQUIRED** :warning:
Hey @marcoshdezcam  & @SigmaSam  :wave:
-
You have done a splendid job with the changes! :tada: Keep it up.
We, however, do have some few more changes to carry out before we proceed (see review comment).
##
## NOTE: These are the variables I used in the tests below: 👇 
```
ARRAY_SIZE = 100
LOWEST_VALUE = 0
HIGHEST_VALUE = 9

describe 'enumerables' do
  let(:array) { Array.new(ARRAY_SIZE) { rand(LOWEST_VALUE...HIGHEST_VALUE) } }
  let(:block) { proc { |num| num < (LOWEST_VALUE + HIGHEST_VALUE) / 2 } }
  let(:words) { %w[dog door rod blade] }
  let(:range) { Range.new(5, 50) }
  let(:hash) { { a: 1, b: 2, c: 3, d: 4, e: 5 } }
  let(:numbers) { [1, 2i, 3.14] }
  let!(:array_clone) { array.clone }
```
## Tests Done: 79
## Failures: 18
- If you have any questions, let me know here or on slack and I will be glad to clarify with you. You can also reach me and connect with me on my social media or email. :eyes:
##
- When you have made the specified changes, please submit another code review. Thank you!
Happy coding :smile:
**[Christian Otieno](https://github.com/ChristianOtieno)**
[Twitter](https://twitter.com/iamchrisotieno) [linkedin](https://www.linkedin.com/in/christianotieno/) [GitLab](https://gitlab.com/christianotieno)
---
*P.S. all requested changes not marked by [OPTIONAL] or are not positive comments are **REQUIRED***

SigmaSam committed:
Fixed my_all to take RegExp.

Commit:
Simplify my_each method. All tests passed

Commit:
Simplify my_each_with_index. All test passed

Commit:
Simplify my_select method. All tests passed

Commit:
FIX: Rubocop issues

SigmaSam committed:
Add escenario for my_all to take, classes and numbers as arguments.

SigmaSam committed:
added tests.rb to git ignore.

SigmaSam committed:
Removed trailing whitespaces.

SigmaSam committed:
Fixed My_None?

SigmaSam committed:
Fixed the issues called by the TSE.

SigmaSam committed:
Removed Rubocop errors.

SigmaSam committed:
Fixed Issues appointed with my_inject, my_all and my_none.

SigmaSam committed:
removed screenshot from README.md

mikenath223 reviewed (changes requested):
### Hi Team 👋

Nice work so far on this. 👌

However, let's fix these issues:

- [ ] my_all when no block or argument is passed and an array is given is expected to return false if at least one of the array elements is false or nil but yours returns true.

- [ ] my_none, when no block or argument is given, should return true if none of the array elements is true, but yours returns false.

- [ ] my_map, when both a block and a proc are given, is expected to execute only the proc and return a result based on the value of running the proc on each element in the array but yours returns an unexpected value.

- [ ] my_inject is expected to be able to search through an array of strings and return the longest word in the string but yours simply returns the array back.

 - [ ] my_inject also when a block is given without an initial value should be able to combine all elements in the array based on the operation specified in the array but yours simply returns the array.

- [ ] my_inject also fails when a range is passed. Also when a symbol is passed as an argument. 

- [ ] my_inject also fails in this example: `[4,6,2,7,9].my_inject(20, :*)`.

- [ ] multiply_els fails to return the right result when tested probably because my_inject isn't working as expected. Please fix it.

Kindly fix these issues and request another review👍
Feel free to contact me on slack(michgolden) if you need further clarification.

HAPPY CODING👨‍💻

mikenath223 commented:
- [ ] In addition, the screenshot on your readme fails to load. Please check it and make amends.

Commit:
Change flow control to match style on my_all, my_any and my_none

Change ternary operator to a modifier if

lekegitrepo commented:
#### Hello Team :clap: :clap:
#### You have done a good job :muscle: However, three test cases are still failing. Check below comments :point_down: 

**_`1. [1, false, 'hi', []].my_all? # => false`_**

**_`2. [nil, false, nil, false].my_none? # => true`_**

**_3. executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum_**
**_`my_proc = { proc { |num| num > 10 } }`_**
**_array.my_map(my_proc) { |num| num < 10 }_**

**_fix these and submit them for review._**

#### _Reviewed by Temitayo Adeleke_

lekegitrepo reviewed (changes requested):
#### Status: Request changes

Commit:
FIX: my_all method 

- Fix when accepting block (Add break statement)
- Fix when no block and no args (Change syntax and delete duplicate declaration)

Commit:
Fix Rubocop Issue on guard clause of my_all && DRY return of my_any?

Commit:
FIX: my_none?

Fix when args empty and block not given case.

SigmaSam committed:
Attended the issues called in the review.

Commit:
Working on my_map

euqueme commented on main.rb:
```diff
@@ -0,0 +1,164 @@
+      end
+    end
+    flag.zero?
+  end
```
- [x] should have working `my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_none?

- [x] when no block or argument is given returns true only if none of the collection members is true
```ruby
[false, nil].my_none? == [false, nil].none? => true
```
- Yours is retrieving the wrong answer
![image](https://user-images.githubusercontent.com/52209067/89251985-b46cc900-d5d5-11ea-9ab3-5d7576a9344e.png)

- Real none?
![image](https://user-images.githubusercontent.com/52209067/89252010-c6e70280-d5d5-11ea-91fa-4926b125b74a.png)

[OPTIONAL]

I think we could get the block of my_none? from my_any? we should only do something like !my_any? for my_none? to work. For example:

  def my_none?(arg = nil, &block)
    !my_any?(arg, &block)
  end

To make this work we have to use block.call(element) instead of yield(element) in my_any?

euqueme commented on main.rb:
```diff
@@ -0,0 +1,164 @@
+    end
+  end
+
+  def my_map(*arg)
+    return to_enum(:my_map) unless block_given?
+
+    new_array = []
+    case block_given?
+    when true && !arg.empty?
+      if arg[0].class == Regexp
+        my_each { |i| yield(i) ? new_array << proc { arg } : nil }
+      else
+        my_each { |i| new_array << i if arg[0].call(i) == true }
+      end
+    when true
+      my_each do |i|
+        new_array << yield(i)
+      end
+    end
+    new_array
+  end
```
* [x] should have working `my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_map 

- [x] in the specifications of the project in the [Odin project](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks) # 12, you were asked to modify your my_map method to take either a proc or a block. And if both a proc and a block are given, only execute the proc. The real map doesn't have this feature, but you were asked to have it so let's work on it

Given 
```ruby
my_proc = Proc.new {|num| num > 10 }
```
Result
```ruby
puts [18, 22, 5, 6] .my_map(my_proc) {|num| num < 10 }  => true true false false
```

euqueme reviewed (changes requested):
> ### Pull Request reviewed by Maria Eugenia Queme P. (TSE)

### Hello @marcoshdezcam and @SigmaSam  :balloon: :lollipop: 

### Thank you for doing the changes requested by the previous reviewers :medal_sports: :medal_sports:  your project is working well and retrieving no errors, but there's a couple of methods that are retrieving the wrong outputs I've put comments directly in your code :point_down: 

### That's it! :raised_hands: 

### Please :pray: do the changes :technologist: :man_technologist: and submit back for a code review when you are done :memo: 

Happy coding! :rocket: 

Any questions? ask me on slack **@ Maru** :woman_technologist: 

PROJECT STATUS: REQUESTED CHANGES :traffic_light:

Commit:
FIX: my_any? to use block.call instead of yield & my_none to use !my_any

Commit:
FIX my_map: Change return of proc from item to "true" || "false"

maelfosso reviewed (changes requested):
# Congratulations :clap: 

Thank you for applying the suggested changes.
You can merge it :+1: 

Good continuation :handshake: 

Thanks
Mael FOSSO
